src/components/Pagination/Nodes/GoToPage.tsx

```diff
@@ -0,0 +1,78 @@
+import { useEffect, useState } from "react";
+
+import { TextInput } from "components/TextInput";
+import { useIsInitialRender } from "utils";
+
+interface GoToPageProps {
+	currentPageIndex: number;
+	delay: number;
+	totalPages: number;
+
+	onPageChange: (
+		e: React.ChangeEvent<HTMLInputElement> | null,
+		newPageIndex: number,
+	) => void;
+
+	// translations
+	pagesText: string;
+	"aria-label": string;
+}
+
+export const GoToPage = ({
+	currentPageIndex,
+	delay,
+	totalPages,
+
+	onPageChange,
+
+	// translations
+	pagesText,
+	"aria-label": ariaLabel,
+}: GoToPageProps) => {
+	const isInitialRender = useIsInitialRender();
+
+	/* NOTE: we need to control both `value` and `event`
+	 * `value` so that we can respond to changes in `currentPageIndex` and the user typing
+	 * `event` so that we can debounce `onPageChange` */
+	const [value, setValue] = useState(currentPageIndex);
+	const [event, setEvent] =
+		useState<React.ChangeEvent<HTMLInputElement> | null>(null);
+
+	useEffect(() => {
+		setValue(currentPageIndex);
+	}, [currentPageIndex]);
+
+	// biome-ignore lint/correctness/useExhaustiveDependencies: only track "real" changes to `newPage`
+	useEffect(() => {
+		const newPage = event ? Number.parseInt(event.target.value, 10) : null;
+
+		// debounce the page change
+		if (!isInitialRender && !!newPage && newPage !== currentPageIndex) {
+			const handler = setTimeout(() => onPageChange(event, newPage), delay);
+			return () => clearTimeout(handler);
+		}
+	}, [event]);
```

Comment:
Below is a typical event handling using debounce.  Your implementation will incur too many state updates and rendering, I am afraid.  
```suggestion
import React, { useState, useCallback } from 'react';
import { debounce } from 'lodash';

const DebouncedInput = () => {
  const [inputValue, setInputValue] = useState("");

  // Create a debounced event handler
  const handleInputChange = useCallback(
    debounce((value) => {
      console.log("Input value:", value);
    }, 300),
    []
  );

  // Attach the debounced handler to the input change event
  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
    handleInputChange(e.target.value);
  };

  return (
    <input
      type="text"
      value={inputValue}
      onChange={onChange}
      placeholder="Type something..."
    />
  );
};

export default DebouncedInput;
```

Reply:
I did some testing and my original way is more performant. Barely, but it is. I suspect that `lodash`'s method does not optimize for React and is thus causing a re-render for every keystroke. 

I'm happy to keep working on this, but I'd like to merge this as-is and create a new ticket for optimization as I've spent quite a bit of time working on this and it is "good enough" IMO.

Re-render test results:
![Screenshot 2024-08-22 at 14 14 18](https://github.com/user-attachments/assets/6f3adaec-3d61-4f81-9085-dc98d61d99cc)

Reply:
I mentioned this above but perhaps we should remove debounce and wait until the user hits enter.

Reply:
> I did some testing and my original way is more performant. Barely, but it is. I suspect that `lodash`'s method does not optimize for React and is thus causing a re-render for every keystroke.
> 
> I'm happy to keep working on this, but I'd like to merge this as-is and create a new ticket for optimization as I've spent quite a bit of time working on this and it is "good enough" IMO.
> 
> Re-render test results: ![Screenshot 2024-08-22 at 14 14 18](https://private-user-images.githubusercontent.com/55896546/360600165-6f3adaec-3d61-4f81-9085-dc98d61d99cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzNTcwOTMsIm5iZiI6MTcyNDM1Njc5MywicGF0aCI6Ii81NTg5NjU0Ni8zNjA2MDAxNjUtNmYzYWRhZWMtM2Q2MS00ZjgxLTkwODUtZGM5OGQ2MWQ5OWNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIyVDE5NTk1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzZTVjNjhjMDQ0ZmU0MzAxMDVhZjdmZjk3MDllMDcxMDJlMTBkOWIxM2QwMWRhZWJhOWUyNzY5YTYyMThiZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DRnOhN0O17kbYZPifg-XS1xefVl3JNNjTmnZmz7GGCg)

A state change will trigger a re render.  You have two state variables here, event and value.  Both change after a keystroke.  In the loadash.debounce version, there is one state variable, value. It changes after each keystroke as well because setValue() is called per stroke.  The performance number does not justify which implementation is better.  But I have never seen event used a dependency in useEffect.  That is a code smell to me.

Reply:
Seeing an `event` dep as a code smell makes sense, but it's what we need to pass along, so it is what needs to be reacted to.

Also, whenever we add a new project dependancy, we're _supposed_ to go through legal to update our license (to give appropriate credit). Thus, I am _extremely_ hesitant to do any work that requires adding a new project dep.

With all of that, I'm leaning towards leaving as is and creating a new story to investigate further. However, per Enrique's comment on this PR, I need to discuss with Matt and Justin if using a debounce is necessary as we may want to simply react to the "Enter" key being pressed.